### PR TITLE
Preprocessing w/o tests

### PR DESCRIPTION
**Pull request by tensionhead**
Author Guidelines
-----------------
- [ ] sadly no, maybe we have to raise this?! ~Is the change set **< 400 lines**?~
This is a dev PR, the points below will be checked once the tests are there
- [ ] Was the code checked for memory leaks/performance bottlenecks?
- [ ] Is the code running locally **and** on the ESI cluster?
- [ ] Is the code running on all supported platforms?

Reviewer Checklist
------------------
- [ ] Are testing routines present?
- [ ] Do **parallel** loops have a set length and correct termination conditions?
- [x] Do objects in the global package namespace perform proper parsing of their input? 
- [x] Do code-blocks provide novel functionality, i.e., no re-factoring using builtin/external packages possible?
- [ ] Code layout
  - [ ] Is the code PEP8 compliant?
  - [ ] Does the code adhere to agreed-upon naming conventions?
  - [ ] Are keywords clearly named and easy to understand?
  - [ ] No commented-out code?
- [ ] Are all docstrings complete and accurate?

### CODE REVIEW

File: syncopy/nwanalysis/ST_compRoutines.py [hunk 1]
```diff
@@ -182,7 +182,6 @@ class ST_CrossSpectra(ComputationalRoutine):
 
     computeFunction = staticmethod(cross_spectra_cF)
 
-    backends = [csd]
```
Comment (joschaschmiedt) [inline syncopy/nwanalysis/ST_compRoutines.py:185]: I suppose these changes to files in the `nwanalysis` folder should not be part of this pull request

Comment (tensionhead) [reply syncopy/nwanalysis/ST_compRoutines.py:185]: Well, yes mr. pedantic ;) On the other hand this just deletes a superfluous line of code and does not change any functionality..

Comment (joschaschmiedt) [reply syncopy/nwanalysis/ST_compRoutines.py:185]: You’re right of course. I started reading and was briefly confused and then had to leave my desk. I‘ll be happy to (properly) review the code in the coming days

Comment (tensionhead) [reply syncopy/nwanalysis/ST_compRoutines.py:185]: Sure, I guess @pantaray does not insist on doing it!

Comment (pantaray) [reply syncopy/nwanalysis/ST_compRoutines.py:185]: Already removed myself from the reviewer list. Please go ahead @joschaschmiedt Thank you!

**Comment by joschaschmiedt** [comment]
I'm curious how much and for what amounts of data this kind of parallel filtering is actually beneficial in terms of performance as filtering is often cheap. The tests may be able to show this.

**Comment by tensionhead** [comment]
According to @KatharineShapcott we can expect extremely long signals (30KHz with recording times up to an hour) on the individual channels, so parallelizing over them might be really beneficial here!

File: syncopy/preproc/firws.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    minus the original low-pass impulse response.
+    In the DSP world the delta distribution corresponds to a mere `1`.
+    """
+    kernel = -kernel
+    # kernel size is always odd
+    kernel[len(kernel) // 2] += 1
+    return kernel
+
+
+def minphaserceps(fkernel):
```
Comment (joschaschmiedt) [inline syncopy/preproc/firws.py:161]: What does the `rceps` part of the function name stand for?

File: syncopy/preproc/firws.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    nSamples = len(fkernel)
+    upsamplingFactor = 1e3  # Impulse response upsampling/zero padding to reduce time-aliasing
+    nFFT = int(2**np.ceil(np.log2(nSamples * upsamplingFactor)))  # Power of 2
+    clipThresh = 1e-8  # -160 dB
+
+    # Spectrum
+    specC = np.abs(np.fft.fft(fkernel, nFFT))
+    specC[specC < clipThresh] = clipThresh  # Clip spectrum to reduce time-aliasing
+
+    # Real cepstrum
```
Comment (joschaschmiedt) [inline syncopy/preproc/firws.py:188]: `cepstrum`?

File: syncopy/preproc/firws.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    # Convolve
+    ires = np.hstack([specR[1:nFFT // 2], 0]) + np.conj(specR[nFFT // 2:nFFT + 1][::-1])
+    specR = np.hstack([
+        specR[0], ires, np.zeros(nFFT // 2 - 2)])
+
+    # Minimum phase
+    MinPhase = np.real(np.fft.ifft(np.exp(np.fft.fft(specR))))
+
+    # Remove zero-padding
+    return MinPhase[:nSamples]
```
Comment (joschaschmiedt) [inline syncopy/preproc/firws.py:200]: I honestly don't understand what the goal of this function is. The provided links where not  helpful for me. Is this a translation of a Fieldtrip function?

**Review by joschaschmiedt — COMMENTED** [review]
Nice and clean groundwork for filtering! I have only two general comments:

- The API of this PR's `preprocessing` is different from Fieldtrip's `ft_preprocessing`. While I like the simpler API here, I think we should discuss this departure from Fieldtrip's API. Also, in the current state the `preprocessing` meta-function covers only filtering, and not other preprocessing steps such as Hilbert transforms, downsampling etc.
- From reading the code in `firws.py` I don't feel confident about judging whether the FIR filtering works as it should. The coming tests will probably cover this.

File: syncopy/preproc/preprocessing.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+                  filter_class='but',
+                  filter_type='lp',
+                  freq=None,
+                  order=None,
+                  direction=None,
+                  window="hamming",
+                  polyremoval=None,
+                  **kwargs
+                  ):
+    """
```
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/preprocessing.py:42]: The FT interface here is really really cumbersome, they have 4 parameters for each option **for each** filter type.. which totals in sth like 50 parameters.. I really don't want to follow them down this rabbit hole. I would suggest we keep the discussion about compatibility around #235. The main functionality of preprocessing is the filtering. Detrending and rectification are trivial features, we can easily add them later. ..and I honestly don't know what a Hilbert transfrom has to do with pre-processing :sweat_smile:, but also this is rather trivial to implement

File: syncopy/preproc/preprocessing.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        raise SPYValueError(legal=lgl, varname="filter_class", actual=filter_class)
+
+    if not isinstance(filter_type, str) or filter_type not in availableFilterTypes:
+        lgl = f"one of {availableFilterTypes}"
+        act = filter_type
+        raise SPYValueError(lgl, 'filter_type', filter_type)
+
+    # check `freq` setting
+    if filter_type in ('lp', 'hp'):
+        scalar_parser(freq, varname='freq', lims=[0, data.samplerate / 2])
```
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/preprocessing.py:105]: I think these are edge cases that no one wants to use anyways right? If you want to filter by the edge-frequencies then we also can't really help you.. also, what would be the sensitive limit then?!

File: syncopy/preproc/preprocessing.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+    # -- here the defaults are filter specific and get set later --
+
+    # filter order
+    if order is not None:
+        scalar_parser(order, varname='order', lims=[0, np.inf], ntype='int_like')
+
+    # check polyremoval
+    if polyremoval is not None:
+        scalar_parser(polyremoval, varname="polyremoval", ntype="int_like", lims=[0, 1])
```
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/preprocessing.py:119]: That's interesting, I recall a discussion with Pascal where he strongly argued that the highest detrending order should anyways be 1..

File: syncopy/preproc/preprocessing.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    check_effective_parameters,
+    check_passed_kwargs
+)
+
+from .compRoutines import But_Filtering, Sinc_Filtering
+
+availableFilters = ('but', 'firws')
+availableFilterTypes = ('lp', 'hp', 'bp', 'bs')
+availableDirections = ('twopass', 'onepass', 'onepass-minphase')
+availableWindows = ("hamming", "hann", "blackmann")
```
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/preprocessing.py:26]: Ah.. thanks!

File: syncopy/preproc/preprocessing.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        if direction is None:
+            direction = 'onepass'
+            msg = f"Setting default direction for FIR filter to '{direction}'"
+            SPYInfo(msg)
+        elif not isinstance(direction, str) or direction not in availableDirections:
+            lgl = "'" + "or '".join(opt + "' " for opt in availableDirections)
+            raise SPYValueError(legal=lgl, varname="direction", actual=direction)
+
+        if order is None:
+            order = int(lenTrials.min())
```
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/preprocessing.py:204]: Yep you are right, I was thinking of the standard 1000 sample trials.. I will change it to either `lenTrials.min()` or 1000 samples

File: syncopy/preproc/preprocessing.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+                                   besides=['filter_class'])
+
+        filterMethod = Sinc_Filtering(samplerate=data.samplerate,
+                                      filter_type=filter_type,
+                                      freq=freq,
+                                      order=order,
+                                      window=window,
+                                      direction=direction,
+                                      polyremoval=polyremoval,
+                                      timeAxis=timeAxis)
```
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/preprocessing.py:221]: Well all our frontends look that way atm.. if you can showcase your factory design in an example we can see if we change that!

File: syncopy/preproc/firws.py (hunk 3 above)
Comment (tensionhead) [reply syncopy/preproc/firws.py:188]: Yap, `cepstrum` is what's  the log-transformed amplitude spectrum is called :man_shrugging:

File: syncopy/preproc/firws.py (hunk 4 above)
Comment (tensionhead) [reply syncopy/preproc/firws.py:200]: Yes, it's a FT translation (as written in the doc string).. this link describes it wonderfully, nothing to add from my side: https://www.katjaas.nl/minimumphase/minimumphase.html

File: syncopy/preproc/compRoutines.py [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+        dat = dat
+
+    # filtering does not change the shape
+    outShape = dat.shape
+    if noCompute:
+        return outShape, np.float32
+
+    # detrend
+    if polyremoval == 0:
+        # SciPy's overwrite_data not working for type='constant' :/
```
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/compRoutines.py:105]: That was a copied comment for weird SciPy behavior.. I will remove it

File: syncopy/preproc/preprocessing.py [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+        of two frequencies for band-stop and band-pass filter.
+    order : int, optional
+        Order of the filter, default is 6.
+        Higher orders yield a sharper transition width
+        or less 'roll off' of the filter, but are more computationally expensive.
+    direction : {'twopass', 'onepass', 'onepasse-minphase'}
+       Filter direction:
+       `'twopass'` - zero-phase forward and reverse filter, IIR and FIR
+       `'onepass'` - forward filter, introduces group delays for IIR, zerophase for FIR
+       `'onepass-minphase' - forward causal/minumum phase filter, FIR only
```
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/preprocessing.py:63]: If you understand the `minphaserceps` function then you know the difference of `onepass-minphase` and `onepass-zerophase` :grin: FIR filter are symmetric, and hence zerophase and acausal.. with the minphase transformation this changes, you get a phase shift but an (almost) causal filter. So the only thing we don't offer atm is the reverse filtering, if ppl really want it we can add it later (this is btw the our default strategy now..)

File: syncopy/preproc/preprocessing.py [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+                  polyremoval=None,
+                  **kwargs
+                  ):
+    """
+    Filtering of time continuous raw data with IIR and FIR filters
+
+    data : `~syncopy.AnalogData`
+        A non-empty Syncopy :class:`~syncopy.AnalogData` object
+    filter_class : {'but', 'firws'}
+        Butterworth (IIR) or windowed sinc (FIR)
```
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/preprocessing.py:48]: Yes good idea, again this PR is to provide the basic functionality.. we can always add some nice stuff later! Maybe open a respective issue?!

File: syncopy/preproc/preprocessing.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (joschaschmiedt) [reply syncopy/preproc/preprocessing.py:42]: I agree that ft_preprocessing is a complex beast that we should probably not mimic. Going with #235, why not call this module `filtering.py` and classify it as *not compatible*?

File: syncopy/preproc/preprocessing.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (joschaschmiedt) [reply syncopy/preproc/preprocessing.py:105]: Good question. But then, why check at all?

File: syncopy/preproc/preprocessing.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (joschaschmiedt) [reply syncopy/preproc/preprocessing.py]: Ok. I'll open an issue about exploring the factory pattern for the meta-functions.

File: syncopy/preproc/preprocessing.py [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+        act = filter_type
+        raise SPYValueError(lgl, 'filter_type', filter_type)
+
+    # check `freq` setting
+    if filter_type in ('lp', 'hp'):
+        scalar_parser(freq, varname='freq', lims=[0, data.samplerate / 2])
+    elif filter_type in ('bp', 'bs'):
+        array_parser(freq, varname='freq', hasinf=False, hasnan=False,
+                     lims=[0, data.samplerate / 2], dims=(2,))
+        freq = np.sort(freq)
```
[the remark this answers is not in the document]
Comment (joschaschmiedt) [reply syncopy/preproc/preprocessing.py:109]: Probably. If the error message is clear, it's easy to fix for the user.

File: syncopy/preproc/preprocessing.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (joschaschmiedt) [reply syncopy/preproc/preprocessing.py:119]: I'd agree with Pascal here ;-)

File: syncopy/preproc/firws.py (hunk 2 above)
Comment (joschaschmiedt) [reply syncopy/preproc/firws.py:161]: Could this be a typo, and it should be `minphaserspec` for minimum phase real spectrum?

File: syncopy/preproc/firws.py (hunk 3 above)
Comment (joschaschmiedt) [reply syncopy/preproc/firws.py:188]: Maybe  typo, and it's actually `spectrum`?

Comment (joschaschmiedt) [reply syncopy/preproc/firws.py:188]: Ok, I now learned about the [Cepstrum](https://en.wikipedia.org/wiki/Cepstrum)

File: syncopy/preproc/firws.py (hunk 2 above)
Comment (joschaschmiedt) [reply syncopy/preproc/firws.py:161]: https://en.wikipedia.org/wiki/Cepstrum

File: syncopy/preproc/firws.py (hunk 4 above)
Comment (joschaschmiedt) [reply syncopy/preproc/firws.py:200]: Ok, I see. I guess I have to find some time to read up about minimum phase filters

**Comment by joschaschmiedt** [comment]
As all code is additional and WIP, I'm ok with merging.
